File: cli/providers.ts
/**
 * Provider API Keys management module.
 */

import { ask, askNumber, pressEnter, clearScreen, confirm } from "./utils/prompt.ts";
import { section, table, success, error, info, colors, color, maskApiKey } from "./utils/display.ts";
import { hasEnvValue, getEnvValue, writeEnvValue, removeEnvValue } from "./utils/env.ts";
import { providers } from "../defaults/providers.ts";

interface ProviderInfo {
  key: string;
  name: string;
  apiKeyName: string;
  hasKey: boolean;
  maskedKey: string;
}

function getProviderList(): ProviderInfo[] {
  return Object.entries(providers).map(([key, provider]) => {
    const apiKeyValue = getEnvValue(provider.apiKeyName);
    return {
      key,
      name: provider.name,
      apiKeyName: provider.apiKeyName,
      hasKey: hasEnvValue(provider.apiKeyName),
      maskedKey: maskApiKey(apiKeyValue),
    };
  });
}

export async function manageProviderKeys(): Promise<void> {
  while (true) {
    clearScreen();
    section("🔑 API Keys de Providers");

    const providerList = getProviderList();

    // Build table
    const headers = ["#", "Provider", "Variable", "Estado"];
    const rows = providerList.map((p, i) => [
      String(i + 1),
      p.name,
      p.apiKeyName,
      p.hasKey
        ? color("✓ " + p.maskedKey, colors.green)
        : color("✗ No configurada", colors.red),
    ]);

    table(headers, rows, [4, 14, 22, 24]);

    console.log();
    info("Selecciona un provider para configurar su API Key");
    console.log();

    const choice = await askNumber("Opción (0 para volver): ");

    if (choice === null || choice === 0) {
      return;
    }

    const selectedProvider = providerList[choice - 1];
    if (!selectedProvider) {
      continue;
    }

    await configureProviderKey(selectedProvider);
  }
}

async function configureProviderKey(provider: ProviderInfo): Promise<void> {
  clearScreen();
  section(`🔑 Configurar ${provider.name}`);

  console.log(`  Variable:  ${provider.apiKeyName}`);
  console.log(`  Estado:    ${provider.hasKey
    ? color("✓ Configurada (" + provider.maskedKey + ")", colors.green)
    : color("✗ No configurada", colors.red)
  }`);
  console.log();

  if (provider.hasKey) {
    console.log("  1. Actualizar API Key");
    console.log("  2. Eliminar API Key");
    console.log("  0. Volver");
    console.log();

    const choice = await askNumber("> ");

    switch (choice) {
      case 1:
        await updateApiKey(provider);
        break;
      case 2:
        await deleteApiKey(provider);
        break;
      default:
        return;
    }
  } else {
    console.log("  1. Agregar API Key");
    console.log("  0. Volver");
    console.log();

    const choice = await askNumber("> ");

    if (choice === 1) {
      await updateApiKey(provider);
    }
  }
}

async function updateApiKey(provider: ProviderInfo): Promise<void> {
  console.log();
  info(`Ingresa la API Key para ${provider.name}`);
  info("(La key se guardará en el archivo .env)");
  console.log();

  const newKey = await ask("API Key: ");

  if (!newKey.trim()) {
    error("API Key vacía, operación cancelada");
    await pressEnter();
    return;
  }

  try {
    await writeEnvValue(provider.apiKeyName, newKey.trim());
    console.log();
    success(`${provider.apiKeyName} guardada correctamente`);
    success(`Provider ${provider.name} ahora disponible`);
  } catch (err) {
    error(`Error guardando la key: ${err}`);
  }

  await pressEnter();
}

async function deleteApiKey(provider: ProviderInfo): Promise<void> {
  console.log();
  const confirmed = await confirm(`¿Eliminar API Key de ${provider.name}?`);

  if (!confirmed) {
    info("Operación cancelada");
    await pressEnter();
    return;
  }

  try {
    await removeEnvValue(provider.apiKeyName);
    console.log();
    success(`${provider.apiKeyName} eliminada`);
    info(`Provider ${provider.name} ya no estará disponible`);
  } catch (err) {
    error(`Error eliminando la key: ${err}`);
  }

  await pressEnter();
}

File: internal/cli/providers.go
package cli

import (
	"fmt"
)

// ManageProviderKeys implements menu option 2: manage provider API keys.
func ManageProviderKeys(_ *App) {
	for {
		ClearScreen()
		Section("🔑 API Keys de Providers")

		list := buildProviderList()

		headers := []string{"#", "Provider", "Variable", "Estado"}
		rows := make([][]string, len(list))
		for i, p := range list {
			var estado string
			if p.hasKey {
				estado = Color(Green, "✓ "+MaskKey(p.value))
			} else {
				estado = Color(Red, "✗ No configurada")
			}
			rows[i] = []string{fmt.Sprintf("%d", i+1), p.name, p.envKey, estado}
		}
		Table(headers, rows, []int{3, 13, 22, 28})

		fmt.Println()
		Info("Selecciona un provider para configurar su API Key")
		fmt.Println()

		n, ok, err := AskNumber("Opción (0 para volver): ")
		if err != nil || !ok || n == 0 {
			return
		}
		if n < 1 || n > len(list) {
			continue
		}
		configureProviderKey(list[n-1])
	}
}

func configureProviderKey(p providerEntry) {
	ClearScreen()
	Section(fmt.Sprintf("🔑 Configurar %s", p.name))

	fmt.Printf("  Variable:  %s\n", p.envKey)
	if p.hasKey {
		fmt.Printf("  Estado:    %s\n", Color(Green, "✓ Configurada ("+MaskKey(p.value)+")"))
	} else {
		fmt.Printf("  Estado:    %s\n", Color(Red, "✗ No configurada"))
	}
	fmt.Println()

	if p.hasKey {
		fmt.Println("  1. Actualizar API Key")
		fmt.Println("  2. Eliminar API Key")
		fmt.Println("  0. Volver")
		fmt.Println()
		n, _, _ := AskNumber("> ")
		switch n {
		case 1:
			updateProviderKey(p)
		case 2:
			deleteProviderKey(p)
		}
	} else {
		fmt.Println("  1. Agregar API Key")
		fmt.Println("  0. Volver")
		fmt.Println()
		n, _, _ := AskNumber("> ")
		if n == 1 {
			updateProviderKey(p)
		}
	}
}

func updateProviderKey(p providerEntry) {
	fmt.Println()
	Info(fmt.Sprintf("Ingresa la API Key para %s", p.name))
	Info("(La key se guardará en el archivo .env)")
	fmt.Println()

	newKey, err := Ask("API Key: ")
	if err != nil || newKey == "" {
		Error("API Key vacía, operación cancelada")
		PressEnter("")
		return
	}

	if err := WriteValue(p.envKey, newKey); err != nil {
		Error(fmt.Sprintf("Error guardando la key: %v", err))
	} else {
		fmt.Println()
		Success(p.envKey + " guardada correctamente")
		Success(fmt.Sprintf("Provider %s ahora disponible", p.name))
	}
	PressEnter("")
}

func deleteProviderKey(p providerEntry) {
	fmt.Println()
	ok, err := Confirm(fmt.Sprintf("¿Eliminar API Key de %s?", p.name), false)
	if err != nil || !ok {
		Info("Operación cancelada")
		PressEnter("")
		return
	}

	if err := RemoveValue(p.envKey); err != nil {
		Error(fmt.Sprintf("Error eliminando la key: %v", err))
	} else {
		fmt.Println()
		Success(p.envKey + " eliminada")
		Info(fmt.Sprintf("Provider %s ya no estará disponible", p.name))
	}
	PressEnter("")
}
